src/main/java/eu/dissco/orchestration/backend/service/MachineAnnotationServiceService.java

```diff
@@ -88,25 +297,75 @@ private void publishUpdateEvent(MachineAnnotationServiceRecord newMasRecord,
           SUBJECT_TYPE);
     } catch (JsonProcessingException e) {
       log.error("Unable to publish message to Kafka", e);
-      rollbackToPreviousVersion(currentMasRecord);
+      rollbackToPreviousVersion(currentMasRecord, true, true);
       throw new ProcessingFailedException("Failed to create new machine annotation service", e);
     }
   }
 
-  private void rollbackToPreviousVersion(MachineAnnotationServiceRecord currentMasRecord) {
+  private void rollbackToPreviousVersion(MachineAnnotationServiceRecord currentMasRecord,
+      boolean rollbackDeployment, boolean rollbackKeda) {
     repository.updateMachineAnnotationService(currentMasRecord);
+    if (rollbackDeployment) {
+      try {
+        deployMasToCluster(currentMasRecord, false);
+      } catch (KubernetesFailedException e) {
+        log.error(
+            "Fatal exception, unable to rollback kubernetes deployment for: {} ",
+            currentMasRecord.pid(), e);
+      }
+    }
+    if (rollbackKeda) {
+      try {
+        updateKedaResource(currentMasRecord, null);
+      } catch (KubernetesFailedException e) {
+        log.error(
+            "Fatal exception, unable to rollback keda for: {} ",
+            currentMasRecord.pid(), e);
+      }
+    }
   }
 
   public void deleteMachineAnnotationService(String id) throws NotFoundException {
     var currentMasOptional = repository.getActiveMachineAnnotationService(id);
     if (currentMasOptional.isPresent()) {
       var deleted = Instant.now();
       repository.deleteMachineAnnotationService(id, deleted);
+      deleteDeployment(currentMasOptional.get());
     } else {
       throw new NotFoundException("Requested machine annotation system: " + id + "does not exist");
     }
   }
 
+  private void deleteDeployment(MachineAnnotationServiceRecord currentMasRecord) {
+    var name = getName(currentMasRecord.pid());
+    try {
+      appsV1Api.deleteNamespacedDeployment(name + DEPLOYMENT,
+          properties.getNamespace(), null, null, null, null, null, null);
+    } catch (ApiException e) {
+      log.error(
+          "Deletion of kubernetes deployment failed for record: {}, with code: {} and message: {}",
+          currentMasRecord, e.getCode(), e.getResponseBody());
+      repository.createMachineAnnotationService(currentMasRecord);
+      throw new ProcessingFailedException("Failed to delete kubernetes resources", e);
+    }
+    try {
+      customObjectsApi.deleteNamespacedCustomObject(KEDA_GROUP, KEDA_VERSION,
+          properties.getNamespace(), KEDA_RESOURCE, name + SCALED_OBJECT, null,
+          null, null, null, null);
+    } catch (ApiException e) {
+      log.error(
+          "Deletion of kubernetes keda failed for record: {}, with code: {} and message: {}",
+          currentMasRecord, e.getCode(), e.getResponseBody());
+      repository.createMachineAnnotationService(currentMasRecord);
```

Comment:
Why do you create a MAS in the repository here?

Reply:
So there was a mas then we tried to delete this but that half-way failed. The mas is still in the cluster so to synchronize db and cluster we re-add the mas to the database.